protocol/state/snapshot.go

```diff
@@ -1,53 +1,96 @@
-// Package state provides types for encapsulating blockchain state.
 package state
 
 import (
+	"fmt"
+
+	"chain/errors"
 	"chain/protocol/bc"
 	"chain/protocol/patricia"
 )
 
 // Snapshot encompasses a snapshot of entire blockchain state. It
-// consists of a patricia state tree and the issuances memory.
+// consists of a patricia state tree and the nonce set.
 //
-// Issuances maps an "issuance hash" to the time (in Unix millis)
-// at which it should expire from the issuance memory.
+// Nonces maps a nonce entry's ID to the time (in Unix millis)
+// at which it should expire from the nonce set.
 type Snapshot struct {
-	Tree      *patricia.Tree
-	Issuances map[bc.Hash]uint64
+	Tree   *patricia.Tree
+	Nonces map[bc.Hash]uint64
 }
 
-// PruneIssuances modifies a Snapshot, removing all issuance hashes
+// PruneNonces modifies a Snapshot, removing all nonce IDs
 // with expiration times earlier than the provided timestamp.
-func (s *Snapshot) PruneIssuances(timestampMS uint64) {
-	for hash, expiryMS := range s.Issuances {
+func (s *Snapshot) PruneNonces(timestampMS uint64) {
+	for hash, expiryMS := range s.Nonces {
 		if timestampMS > expiryMS {
-			delete(s.Issuances, hash)
+			delete(s.Nonces, hash)
 		}
 	}
 }
 
 // Copy makes a copy of provided snapshot. Copying a snapshot is an
-// O(n) operation where n is the number of issuance hashes in the
-// snapshot's issuance memory.
+// O(n) operation where n is the number of nonces in the
+// snapshot's nonce set.
 func Copy(original *Snapshot) *Snapshot {
 	// TODO(kr): consider making type Snapshot truly immutable.
 	// We already handle it that way in many places (with explicit
 	// calls to Copy to get the right behavior).
 	c := &Snapshot{
-		Tree:      new(patricia.Tree),
-		Issuances: make(map[bc.Hash]uint64, len(original.Issuances)),
+		Tree:   new(patricia.Tree),
+		Nonces: make(map[bc.Hash]uint64, len(original.Nonces)),
 	}
 	*c.Tree = *original.Tree
-	for k, v := range original.Issuances {
-		c.Issuances[k] = v
+	for k, v := range original.Nonces {
+		c.Nonces[k] = v
 	}
 	return c
 }
 
 // Empty returns an empty state snapshot.
 func Empty() *Snapshot {
 	return &Snapshot{
-		Tree:      new(patricia.Tree),
-		Issuances: make(map[bc.Hash]uint64),
+		Tree:   new(patricia.Tree),
+		Nonces: make(map[bc.Hash]uint64),
+	}
+}
+
+// ApplyBlock updates s in place.
+func (s *Snapshot) ApplyBlock(block *bc.BlockEntries) error {
```

Comment:
Given the current (and IMO gratifyingly clear) structure of the code, I think they should. Mutating Snapshots is no more than what the code was already doing. But I expanded on the TODO comment to point out that the `Apply` functions (and `PrunceNonces`) will have to change if Snapshots become immutable.